crates/subspace-service/src/domains/request_handler.rs

```diff
@@ -0,0 +1,248 @@
+        format!(
+            "/{}/{}/last-confirmed-block/1",
+            array_bytes::bytes2hex("", genesis_hash),
+            fork_id
+        )
+    } else {
+        format!(
+            "/{}/last-confirmed-block/1",
+            array_bytes::bytes2hex("", genesis_hash)
+        )
+    }
+}
+
+/// Request last confirmed domain block data from a peer.
+#[derive(Clone, PartialEq, Encode, Decode, Debug)]
+pub struct LastConfirmedBlockRequest<Block: BlockT> {
+    pub domain_id: DomainId,
+    pub block_hash: Option<Block::Hash>,
+}
+
+#[derive(Clone, PartialEq, Encode, Decode, Debug)]
+pub struct LastConfirmedBlockResponse<Block: BlockT, DomainHeader: Header> {
+    pub last_confirmed_block_receipt: ExecutionReceiptFor<DomainHeader, Block, Balance>,
+}
+
+/// Handler for incoming block requests from a remote peer.
+pub struct LastDomainBlockERRequestHandler<Block: BlockT, Client, DomainHeader> {
+    request_receiver: async_channel::Receiver<IncomingRequest>,
+
+    _phantom: PhantomData<(Block, DomainHeader)>,
+
+    client: Arc<Client>,
+}
+
+impl<Block, Client, DomainHeader> LastDomainBlockERRequestHandler<Block, Client, DomainHeader>
+where
+    Block: BlockT,
+    Client: ProvideRuntimeApi<Block>
+        + BlockBackend<Block>
+        + ProofProvider<Block>
+        + HeaderBackend<Block>
+        + Send
+        + Sync
+        + 'static,
+    Client::Api: DomainsApi<Block, DomainHeader> + FraudProofApi<Block, DomainHeader>,
+    DomainHeader: Header,
+{
+    /// Create a new [`LastDomainBlockERRequestHandler`].
+    pub fn new<NB>(
+        protocol_id: &ProtocolId,
+        fork_id: Option<&str>,
+        client: Arc<Client>,
+        num_peer_hint: usize,
+    ) -> (Self, NB::RequestResponseProtocolConfig)
+    where
+        NB: NetworkBackend<Block, <Block as BlockT>::Hash>,
+    {
+        // Reserve enough request slots for one request per peer when we are at the maximum
+        // number of peers.
+        let capacity = std::cmp::max(num_peer_hint, 1);
+        let (tx, request_receiver) = async_channel::bounded(capacity);
+
+        let protocol_config = generate_protocol_config::<_, Block, NB>(
+            protocol_id,
+            client
+                .block_hash(0u32.into())
+                .ok()
+                .flatten()
+                .expect("Genesis block exists; qed"),
+            fork_id,
+            tx,
+        );
+
+        (
+            Self {
+                request_receiver,
+                client,
+                _phantom: PhantomData,
+            },
+            protocol_config,
+        )
+    }
+
+    /// Run [`StateRequestHandler`].
+    pub async fn run(mut self) {
+        while let Some(request) = self.request_receiver.next().await {
+            let IncomingRequest {
+                peer,
+                payload,
+                pending_response,
+            } = request;
+
+            match self.handle_request(payload, pending_response, &peer) {
```

Comment:
We may need to check if the node is in major sync or not before handling the request, because for the nodes that are still in major sync their `LastDomainBlockER` is stale. Or perhaps this is already checked in the networking stack?